#183 Core skills page (#186)
sparkfabrik/company-playbook: content/procedures/assessing-core-skills.md

The procedure is easy:

* Use the guidelines and checklists below to assess the necessary skill level.
* If the degree of confidence is low (most of the checklist elements are unknown or uncertain), the whole module is worth studying from scratch. If most of the topics are known and just a few ancillary are missing (ex. the assessee never used `sed` and `awk` but for the rest she's confident with a Unix shell), than the module may be skipped.
* Take a list of the modules that have to be studied.
* Head to the [Plan the training on Core Skills](#plan-the-training-on-core-skills) section below to schedule the training.

## Assess Core Skills proficiency

Follow this section, module by module.

### Basics

#### Command line / Shell

Software Developers should have a basic understanding of:

* Files and folder management:
  * **Navigation**: `ls`, `cd`, glob patterns
  * **Search**: `find`, `which`
  * **File management**: `cat`, `less`/`more`, `head`/`tail`, `touch`
  * **Permissions**: `chmod` and `chown`, permissions sets, file attributes

> * **Shell configuration**: `.profile`/`.zprofile`, ``.bashrc`/`.zshrc`, environment variables, path and command scope, etc

Reviewer: extra backtick before .bashrc

Author: Ouch, thanks!

* **Make/Makefile**: targets, dependencies, recipes, rules

Senior Software Developers and Cloud Native Engineers should be confident with the above, plus having a decent understanding of:

* **Scripting**: Variables, arguments, cycles, conditions, functions, echoing
* **Text processing**: `grep`, `sed`, `awk`

#### Security

Software Developers should have a basic understanding of:

* Asymmetric cryptography, SSH, authentication, signatures
* Best-practices for secrets-management in software repo (environment variables and `.env` files, `.gitignore`, `.dockerignore`)
* Security monitoring tools: CVE, GitHub advisory database, Dependabot / Renovate

Senior Software Developers and Cloud Native Engineers should have a clear understanding about:

* GPG, difference between authentication and authorization, OAuth framework, hashing
* Best-practices for the development of secure software

#### Networking

Software Developers should have a basic understanding of:

* Interfaces and ports, what's `localhost`, how domain names resolution works, HTTP basics, how HTTPS works.

Senior Software Developers and Cloud Native Engineers should have a clear understanding about:

> * TCP/UDP-IP protocol, the concept of application-level protocols, DNS configuration basics (authoritative zone, record types, TTL and propagation), local domains resolution.

Reviewer 2: I would compress the "TCP/UDP-IP protocol, the concept of application-level protocols" using "Network Stack".

It is only because the "TCP/UDP-IP" is not totally clear to me.

Author: Sounds right, what about "How the network stack works, the different levels of protocols, DNS configuration basics, ..." etc ?

Author: Changed in last commit!

### Tools

#### YAML

Everyone should have a basic understanding of:

* Basic knowledge of the use-cases and syntax
* Knowledge of features like comments or references

#### Git

Everyone should have a basic understanding of:

* **Basic usage**: status, staging, commit, tag, reset, history / log, branch, remotes, merge, rebase, cherry picking, push, pull
* **Concepts**: what's a repository, the working tree, the `.gitignore` file, how to diff and move along the commit history

> * **Workflow**: different branching models (git-flow, GitHub or Gitlab flow), pull/merge request, peer review, squashing commits, rebase and merge, merge commits, linear history

Reviewer 2: ```suggestion
* **Workflow**: different branching models (git-flow, GitHub or GitLab flow), pull/merge request, peer review, squashing commits, rebase and merge, merge commits, linear history
```

Just for consistency

* **Conventions**: understand the relation between workflow-related conventions (commit granularity, versioning, commit messages format, branching model, etc) and the working context (social contracts, environments, cadences, etc)

#### GitHub Copilot

That's a "bonus track". Developers are the most impacted by the usage of Copilot, so just ask if they know what it is and if they every tried it.

### DevOps

#### Docker and docker-compose

Software Developers should have a basic understanding of:

> * **Concepts**: Docker daemon and client, general anatomy of a Dockerfile, images, tags, digest, containers, volumes, registries, and what is DockerHub

Reviewer 2: ```suggestion
* **Concepts**: Docker daemon and client, general anatomy of a Dockerfile, images, tags, digest, containers, volumes, registries, and what is a container registry
```

I would prefer to use the generic technology instead of the specific product name. We also use other container registries, like the GitHub Container Registry, more than Docker Hub.

Author: yep, but "registries" already appear right before "what is dockerhub". The idea here is "you know that there is a public registry and how it's positioned in the global ecosystem".

I propose to rephrase the whole line as follows:

`* **Concepts**: Docker daemon and client, general anatomy of a Dockerfile, images, tags, digest, containers, volumes, CONTAINER registries, and what is DockerHub`

Is this any good?

* **Basic commands**: pull, push, run, stop, ls, rm, ps, prune
* **Docker-Compose**: 
  * Understanding declarative approach

>   * Anatomy of the `docker-compose.yml` file (services, build, volumes, ports, etc)

Reviewer 2: ```suggestion
  * Anatomy of the `docker-compose.yml` file (services, build, volumes, ports, environment variables interpolation, etc)
```

  * Basic commands (up, down, logs) and detachment

Senior Software Developers and Cloud Native Engineers should be confident with the above, plus having a more specific understanding of:

> * **Dockerfiles**: levels, caching, basic commands (FROM, COPY, ADD, RUN, ENTRYPOINT, CMD, ENV, ARG), multi stage builds

Reviewer 2: ```suggestion
* **Dockerfiles**: levels, caching, basic commands (FROM, COPY, ADD, RUN, ENTRYPOINT, CMD, ENV, ARG), multi stage and multi arch builds
```

I don't know if this is a skill that the Seniors must have, but I think that it is should be added also because we have two architectures (arm and amd64 at least) in our stack.
